hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/HybridReadDeciderActor.scala

```diff
@@ -1,43 +1,86 @@
 package cromwell.services.metadata.hybridcarbonite
 
-import akka.actor.{Actor, ActorRef, PoisonPill, Props}
-import cromwell.services.metadata.MetadataService.{MetadataReadAction, MetadataServiceResponse}
+import akka.actor.FSM.Normal
+import akka.actor.{ActorRef, FSM, LoggingFSM, Props}
+import cromwell.services.metadata.MetadataService.{MetadataReadAction, MetadataServiceResponse, QueryForWorkflowsMatchingParameters, WorkflowMetadataReadAction, WorkflowQueryFailure, WorkflowQuerySuccess}
+import cromwell.services.metadata.hybridcarbonite.HybridReadDeciderActor._
 import cromwell.services.metadata.impl.builder.MetadataBuilderActor.FailedMetadataResponse
+import cromwell.services.metadata.{MetadataArchiveStatus, WorkflowQueryKey}
 
 import scala.concurrent.{ExecutionContext, Future}
-import scala.util.{Failure, Success}
 
-class HybridReadDeciderActor(classicMetadataServiceActor: ActorRef, carboniteMetadataServiceActor: ActorRef) extends Actor {
+class HybridReadDeciderActor(classicMetadataServiceActor: ActorRef, carboniteMetadataServiceActor: ActorRef) extends LoggingFSM[HybridReadDeciderState, HybridReadDeciderData] {
+
+  startWith(Pending, NoData)
 
   implicit val ec: ExecutionContext = context.dispatcher
 
-  // TODO: [CARBONITE] Decide which actor to send the read request to
-  // NOTE: the 'Future' return value was arbitrary to demonstrate the concept... feel free to refactor all this logic into something better
-  //  like (eg) sending a message to the summary table followed by some appropriate forwarding action when the response comes back
-  def decide(read: MetadataReadAction): Future[ActorRef] = Future.successful(classicMetadataServiceActor)
-
-  var sndr: Option[ActorRef] = None
-
-  override def receive: Receive = {
-    case read: MetadataReadAction =>
-      sndr = Option(sender())
-      decide(read) onComplete {
-        case Success(forwardTo) =>
-          forwardTo ! read
-        case Failure(e) =>
-          respondAndStop(FailedMetadataResponse(read, new Exception("Failed to decide which metadata service to forward request to", e)))
+  when(Pending) {
+    case Event(read: MetadataReadAction, NoData) =>
+      val sndr = sender()
+
+      read match {
+        case wmra: WorkflowMetadataReadAction =>
+          classicMetadataServiceActor ! QueryForWorkflowsMatchingParameters(Vector(WorkflowQueryKey.Id.name -> wmra.workflowId.toString))
+          goto(RequestingMetadataArchiveStatus) using WorkingData(sndr, read)
+
+        case query: QueryForWorkflowsMatchingParameters =>
+          classicMetadataServiceActor ! query
+          goto(WaitingForMetadataResponse) using WorkingData(sndr, read)
+      }
+  }
+
+  when(RequestingMetadataArchiveStatus) {
+    case Event(WorkflowQuerySuccess(response, _), wd: WorkingData) =>
+      if (response.results.size == 1 && response.results.headOption.exists(_.metadataArchiveStatus == MetadataArchiveStatus.Archived)) {
```

Comment:
Is that size check necessary? And if not can the ```response.results.headOption.exists(_.metadataArchiveStatus == MetadataArchiveStatus.Archived))``` just be part of the match?

Reply:
👍 I'll break this into separate a `0` case (which probably indicates "hasn't been summarized yet") and `> 1` case (which probably indicates a horrible, horrible error)